### PR TITLE
app/health: implement health checks

### PR DESCRIPTION
**Pull request by corverroos**
PoC of Health Checks V2. See `checks.go`

category: feature 
ticket: none

### CODE REVIEW

File: app/health/checks.go [hunk 1]
```diff
@@ -0,0 +1,100 @@
...
+		Severity:    severityCritical,
+		Func: func(s store, m metadata) bool {
+			counts := s.Get("p2p_peer_connection_total", countNonZeroLabels)
+			return gaugeMin(counts) < float64(m.QuorumPeers)
+		},
+	},
+	{
+		Name:        "pending_validators",
+		Description: "Pending validators detected. Activate them to start validating.",
+		Type:        typeNegative,
```
Comment (xenowits) [inline app/health/checks.go:82]: would we still call this negative if there are no pending validators, ie, `gaugeMax(counts) == 0` ?

**Review by xenowits — APPROVED** [review]

File: app/health/select.go [hunk 2]
```diff
@@ -0,0 +1,137 @@
...
+			val = metric.Counter.GetValue()
+		case pb.MetricType_GAUGE:
+			val = metric.Gauge.GetValue()
+		default:
+			panic("invalid metric type for simple value labelSelector")
+		}
+
+		if val > max.Counter.GetValue() {
+			max = metric
+		}
```
Comment (dB2510) [inline app/health/select.go:32]: how do you know that `max` metric will always be a counter metric? i think it depends on `metric` variable of the for loop which gets assigned to max in this if block

Comment (dB2510) [reply app/health/select.go:32]: maybe it is better to have two variables: `maxValue` and `maxMetric`. `maxValue` compares the values of the metrics and `maxMetric` stores the max metric based on the value.

File: app/health/select.go [hunk 3]
```diff
@@ -0,0 +1,137 @@
...
+
+func countNonZeroLabels(metricsFam *pb.MetricFamily) *pb.Metric {
+	gauge := &pb.Metric{
+		Gauge:       new(pb.Gauge),
+		TimestampMs: metricsFam.Metric[0].TimestampMs,
+	}
+
+	for _, metric := range metricsFam.Metric {
+		if metric.Gauge.GetValue() != 0 {
+			*guage.Gauge.Value++
```
Comment (dB2510) [inline app/health/select.go:46]:
```suggestion
			*gauge.Gauge.Value++
```

File: app/health/select.go [hunk 4]
```diff
@@ -0,0 +1,137 @@
...
+			panic("histogram sum not implemented yet")
+		default:
+			panic("invalid metric type for simple value labelSelector")
+		}
+	}
+
+	return sum
+}
+
+func l(name, val string) *pb.LabelPair {
```
Comment (dB2510) [inline app/health/select.go:79]: can we have a proper name of this function?

**Review by dB2510 — COMMENTED** [review]
the approach looks good to me, looking forward to how it will interact with other components.

File: app/health/reducers.go [hunk 5]
```diff
@@ -0,0 +1,48 @@
+// Copyright © 2022-2023 Obol Labs Inc. Licensed under the terms of a Business Source License 1.1
+
+package health
+
+import pb "github.com/prometheus/client_model/go"
```
[the remark this answers is not in the document]
Comment (corverroos) [reply app/health/reducers.go]: it is protobufs, so pb is common pattern

File: app/health/select.go (hunk 2 above)
Comment (corverroos) [reply app/health/select.go:35]: updated

File: app/health/select.go (hunk 3 above)
Comment (corverroos) [reply app/health/select.go]: fixed

File: app/health/select.go (hunk 4 above)
Comment (corverroos) [reply app/health/select.go]: no 😄

**Review by gsora — APPROVED** [review]

File: app/health/checker.go [hunk 6]
```diff
@@ -0,0 +1,126 @@
...
+	checks       []check
+	metrics      [][]*pb.MetricFamily
+	gatherer     prometheus.Gatherer
+	scrapePeriod time.Duration
+	maxScrapes   int
+	logFilter    z.Field
+}
+
+// Run runs the health checker until the context is canceled.
+func (c *Checker) Run(ctx context.Context) {
```
Comment (dB2510) [inline app/health/checker.go:49]: i assume this will be a lifecycle hook in app package later.

File: app/health/checker.go [hunk 7]
```diff
@@ -0,0 +1,126 @@
...
+				log.Warn(ctx, "Failed to scrape metrics", err)
+				continue
+			}
+
+			c.instrument(ctx)
+		}
+	}
+}
+
+// Instrument runs all health checks and updates the check gauge.
```
Comment (dB2510) [inline app/health/checker.go:68]:
```suggestion
// instrument runs all health checks and updates the check gauge.
```

**Review by xenowits — APPROVED** [review]

**Review by dB2510 — APPROVED** [review]

**Pull request merged**